Add blog post about Google Summer of Code 2024 project results (#1419)
rust-lang/blog.rust-lang.org: posts/2024-11-01-gsoc-2024-results.md

Last, but not least, Eitaro also created a completely new user interface for the benchmark suite, which runs entirely in the [terminal](https://github.com/rust-lang/rustc-perf/pull/1955). Using this interface, Rust compiler contributors can examine the performance of the compiler without having to start the rustc-perf website, which can be challenging to deploy locally.

Apart from the mentioned contributions, Eitaro also made a lot of other smaller improvements to various parts of the benchmark suite. Thank you for all your work!

### Move cargo shell completions to Rust

- Contributor: [shanmu](https://github.com/shannmu)
- Mentor: [Ed Page](https://github.com/epage)
- [Final report](https://hackmd.io/@PthRWaPvSmS_2Yu_GLbGpg/Hk-ficKpC)

> TODO

Reviewer: Cargo's completion scripts have been hand maintained and frequently broken when changed.  The goal for this effort was to have the completions automatically generated from the definition of Cargo's command-line with extension points for the dynamically generated results.

shanmu took the prototype for clap's dynamic completions and got it working and tested for common shells as well as extended the parser to cover more cases.  They then added extension points for CLI's to provide custom completion results.

In the next phase, shanmu added this to nightly Cargo and added different custom completers to match what the hand written completions do.

See the [instructions](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#native-completions) for trying this out!  That also lists where you can provide feedback.

For whats left before this can be stabilized, see
- [clap#3166](https://github.com/clap-rs/clap/issues/3166)
- [cargo#14520](https://github.com/rust-lang/cargo/issues/14520)

Reviewer 2: Should we link individuals to their GitHub handles?

Author: Thanks a lot! I did some brief editing on this and also added and example, so that people see how cool this is :sunglasses: Let me know what do you think. https://github.com/rust-lang/blog.rust-lang.org/pull/1419/commits/468602997358a5c8e90e49fc16214399d1689d03

Author: > Should we link individuals to their GitHub handles?

I added a link to shanmu (plus some tool links), but I don't think that we need to repeat the links everywhere.

Reviewer: Yeah, I left the link off because it was already in the "Contributor" line but don't mind if it also appears in the body.

### Rewriting Esoteric, Error-Prone Makefile Tests Using Robust Rust Features

- Contributor: [Julien Robert](https://github.com/Oneirical)
- Mentor: [Jieyou Xu](https://github.com/jieyouxu)
- [Final report](https://oneirical.github.io/gsocfinal/)

> TODO

Reviewer 3: The [`run-make` test suite][run-make] previously had tests based on `Makefile`s. However, this setup [posed two primary problems][initial-run-make-issue]:

1. You could only run the test suite under `windows-gnu` (e.g. MSYS2) and not natively under `windows-msvc`. This is problematic because `x86_64-pc-windows-msvc` is also a Tier 1 target, yet it doesn't receive native test coverage from the `run-make` test suite. Contributors also had to perform additional setup like MSYS2 to be able to run the test suite at all.
2. Some of the `Makefile` syntax are quite esoteric, with errors easy to silently ignore and mistakes very difficult to spot, even to multiple reviewers.

Julien helped to convert the `Makefile`-based `run-make` tests into plain Rust-based tests, supported by a test support library called `run_make_support`. It was not a trivial "rewrite this in Rust" kind of deal. In this project, Julien:

- Significantly improved the test documentation;
- Fixed multiple bugs that were present in the `Makefile` versions that had gone unnoticed for *years* -- some tests were never testing anything or silently ignored failures, so even if the subject being tested regressed, these tests would not have caught that.
- Added to and improved the test support library API and implementation; and
- Improved code organization within the tests to make them easier to understand and maintain.

As a result, I believe Julien has significantly improved the robustness of the `run-make` test suite, and improved the ergonomics of modifying existing `run-make` tests and authoring new `run-make` tests. Multiple contributors have expressed that they were more willing to work with the Rust-based `run-make` tests over the `Makefile` versions.

[run-make]: https://github.com/rust-lang/rust/tree/master/tests/run-make
[initial-run-make-issue]: https://github.com/rust-lang/rust/issues/40713

Reviewer 3: (Feel free to compress this to edit as suitable)

Author: Thanks a lot! I tried to add some initial context, just to let people know what is the run-make test suite, and also did some editing. Let me know what do you think.

Reviewer 3: Sounds good to me, thanks!

quite happy with the GSoC program and that they had learned a lot. We are of course also incredibly grateful for all their
contributions - some of them have even continued contributing after their project has ended, which is really awesome.
In general, we think that Google Summer of Code 2024 was a success for the Rust Project, and we are looking forward to
participating in GSoC (or similar programs) again in the near future. If you are interested in becoming a (GSoC) contributor,
check out our [project idea list](https://github.com/rust-lang/google-summer-of-code).

Below you can find a brief summary of each of our GSoC 2024 projects, including feedback
from the contributors and mentors themselves. You can find more information about the projects
[here][gsoc-project-list].

Max:
"The interaction with the community and the other GSoC contributors was really great, and I think encouraging that even more would be even better for next year."

### Adding lint-level configuration to cargo-semver-checks

- Contributor: [Max Carr](https://github.com/suaviloquence/)
- Mentor: [Predrag Gruevski](https://github.com/obi1kenobi)
- [Final report](https://blog.mcarr.one/gsoc-final/)

> The goal of this project was to add the ability to configure lint levels to [cargo-semver-checks](https://github.com/obi1kenobi/cargo-semver-checks), a tool designed for automatically detecting semantic versioning conflicts, which might one day become a part of Cargo itself. Max has achieved this goal by implementing a [comprehensive system](https://github.com/obi1kenobi/cargo-semver-checks?tab=readme-ov-file#lint-level-configuration) for configuring `cargo-semver-checks` lints directly in the `Cargo.toml` manifest file. He has also extensively discussed the design with the Cargo team, to ensure that it is compatible with how other Cargo lints are configured.

Reviewer 4: The first sentence here is very long.

Suggestion with a reworked first sentence and a few other minor edits:
```suggestion
[cargo-semver-checks](https://github.com/obi1kenobi/cargo-semver-checks) is a tool designed for automatically detecting semantic versioning conflicts, which is planned to one day become a part of Cargo itself. The goal of this project was to enable cargo-semver-checks to ship additional *opt-in* lints by allowing users to configure which lints run in which cases, and whether their findings are reported as errors or warnings. Max achieved this goal by implementing a [comprehensive system](https://github.com/obi1kenobi/cargo-semver-checks?tab=readme-ov-file#lint-level-configuration) for configuring `cargo-semver-checks` lints directly in the `Cargo.toml` manifest file. He also extensively discussed the design with the Cargo team to ensure that it is compatible with how other Cargo lints are configured, and won't present a future compatibility problem for merging cargo-semver-checks into Cargo.
```

Author: Thanks! Applied the change.

Even though this project did not fulfill its original goals and more work will be needed in the future to get a functional prototype of WebAssembly proc macros, we are thankful for Apurva's contributions. The WebAssembly loading prototype is a good start, and Apurva's exploration of proc macro internals should serve as a useful reference for anyone working on this feature in the future. Going forward, we will try to describe more incremental steps for our GSoC projects, as this project was perhaps too ambitious from the start.

### Tokio async support in Miri

- Contributor: [Tiffany Pek Yuan](https://github.com/tiif)
- Mentor: [Oli Scherer](https://github.com/oli-obk)
- [Final report](https://gist.github.com/tiif/3e08ba6e8cfb1d078e6155410108ae48)

> TODO

Reviewer 5: ```suggestion
Running programs using `tokio` under miri to check for undefined behaviour failed for all nontrivial progrograms (so those that ever `await` on anything) due to a fundamental missing miri feature: `epoll` support on linux, and similar APIs on the other major platforms.

Tiffany implemented the basic `epoll` operations needed to cover the majority of the tokio test suite by crafting pure `libc` code examples that exercised those `epoll` operations, and then implementing their emulation in miri itself. At times this required refactoring core miri components like file description handling, as they were created solely with blocking operations in mind.

Suprising to everyone (though probably not tokio-internals experts), once these core `epoll` operations were finished, operations like async file reading and writing worked out of the box. Due to limitations of non-blocking file operations, tokio wrapped those file operations in dedicated threads, which was already supported by miri.

The project finished successfully (and with strech goals like async file oporations reached), Tiffany proceeded to contact tokio maintainers and worked with them to run miri on most tokio tests in CI. Good news: so far no soundness problems have been discovered! Tiffany is now a regular contributor to miri, with a focus on continuing to expand the set of supported nonblocking file descriptor operations.
```

[miri][miri] is an intepreter that can find possible instances of undefined behavior in your Rust code, which is being used across the Rust ecosystem. However, it was not possible to run it on any non-trivial programs (those that ever `await` on anything) using [tokio][tokio], due a to a fundamental missing feature: support for the `epoll` syscall on Linux (and similar APIs on other major platforms).

Tiffany implemented the basic `epoll` operations needed to cover the majority of the tokio test suite, by crafting pure `libc` code examples that exercised those `epoll` operations, and then implementing their emulation in miri itself. At times, this required refactoring core miri components like file description handling, as they were originally created solely with blocking operations in mind.

Suprising to everyone (though probably not tokio-internals experts), once these core `epoll` operations were finished, operations like async file reading and writing started working out of the box! Due to limitations of non-blocking file operations offered by operating systems, tokio wrapped those file operations in dedicated threads, which was already supported by miri.

> Once Tiffany has finished the project, including stretch goals like implementing async file operations, she proceeded to contact tokio maintainers and worked with them to run miri on most tokio tests in CI. And we have good news: so far no soundness problems have been discovered! Tiffany become a regular contributor to miri, focusing on continuing to expand the set of supported non-blocking file descriptor operations. We thank her for all her contributions!

Reviewer 5: ```suggestion
Once Tiffany has finished the project, including stretch goals like implementing async file operations, she proceeded to contact tokio maintainers and worked with them to run miri on most tokio tests in CI. And we have good news: so far no soundness problems have been discovered! Tiffany has become a regular contributor to miri, focusing on continuing to expand the set of supported non-blocking file descriptor operations. We thank her for all her contributions!
```

> Tiffany implemented the basic `epoll` operations needed to cover the majority of the tokio test suite, by crafting pure `libc` code examples that exercised those `epoll` operations, and then implementing their emulation in miri itself. At times, this required refactoring core miri components like file description handling, as they were originally created solely with blocking operations in mind.
> 
> Suprising to everyone (though probably not tokio-internals experts), once these core `epoll` operations were finished, operations like async file reading and writing started working out of the box! Due to limitations of non-blocking file operations offered by operating systems, tokio wrapped those file operations in dedicated threads, which was already supported by miri.
> 
> Once Tiffany has finished the project, including stretch goals like implementing async file operations, she proceeded to contact tokio maintainers and worked with them to run miri on most tokio tests in CI. And we have good news: so far no soundness problems have been discovered! Tiffany has become a regular contributor to miri, focusing on continuing to expand the set of supported non-blocking file descriptor operations. We thank her for all her contributions!

Reviewer 6: Thanks for the writeup, just a few nitpicks:

I remember file description handling is refactored to support blocking operation for eventfd and friends. 

```suggestion
Tiffany implemented the basic `epoll` operations needed to cover the majority of the tokio test suite, by crafting pure `libc` code examples that exercised those `epoll` operations, and then implementing their emulation in miri itself. At times, this required refactoring core miri components like file description handling, as they were originally created solely with non-blocking operations in mind.

Suprising to everyone (though probably not tokio-internals experts), once these core `epoll` operations were finished, operations like async file reading and writing started working out of the box! Due to limitations of non-blocking file operations offered by operating systems, tokio wrapped those file operations in dedicated threads, which was already supported by miri.

Once Tiffany has finished the project, including stretch goals like implementing async file operations, she proceeded to contact tokio maintainers and worked with them to run miri on most tokio tests in CI. And we have good news: so far no soundness problems have been discovered! Tiffany has become a regular contributor to miri, focusing on continuing to expand the set of supported blocking file descriptor operations. We thank her for all her contributions!
```

Author: Thanks! Did some modifications.

Predrag, who is the author of `cargo-semver-checks` and who mentored Max on this project, was very happy with his contributions that even went beyond his original project scope:

> He designed and built one of our most-requested features, and produced design prototypes of several more features our users would love. He also observed that writing quality CLI and functional tests was hard, so he overhauled our test system to make better tests easier to make. Future work on cargo-semver-checks will be much easier thanks to the work Max put in this summer.

Great work, Max!

> ### Implementation of a Faster Register Allocator For Cranelift

Reviewer 7: ```suggestion
### Implementation of a faster register allocator for Cranelift
```

[shanmu](https://github.com/shannmu) took the prototype for dynamic completions in [clap][clap] (the command-line parser used by Cargo), got it working and tested for common shells, as well as extended the parser to cover more cases. They then added extension points for CLI's to provide custom completion results that can be generated on the fly.

In the next phase, shanmu added this to nightly Cargo and added different custom completers to match what the handwritten completions do. As an example, with this feature enabled, when you type `cargo test --test=` and hit the Tab key, your shell will autocomplete all the test targets in your current Rust crate! If you are interested, see the [instructions][cargo-shell-completion] for trying this out! The link also lists where you can provide feedback.

You can also check out the following issues to find out what is left before this can be stabilized:
- [clap#3166](https://github.com/clap-rs/clap/issues/3166)
- [cargo#14520](https://github.com/rust-lang/cargo/issues/14520)

[clap]: https://github.com/clap-rs/clap
[cargo-shell-completion]: https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#native-completions

> ### Rewriting Esoteric, Error-Prone Makefile Tests Using Robust Rust Features

Reviewer 7: ```suggestion
### Rewriting esoteric, error-prone makefile tests using robust Rust features
```

### Rewriting the Rewrite trait

- Contributor: [SeoYoung Lee](https://github.com/ding-young)
- Mentor: [Yacin Tmimi](https://github.com/ytmimi)
- [Final report](https://ding-young.github.io/posts/gsoc-final/)

[rustfmt] is a Rust code formatter that is widely used across the Rust ecosystem thanks to its direct integration within Cargo. Usually, you just run `cargo fmt` and you can immediately enjoy a properly formatted Rust project. However, there are edge cases in which `rustfmt` can fail to format your code. That is not such an issue on its own, but it becomes more problematic when it fails *silently*, without giving the user any context about what went wrong. This is what sometimes happens in `rustfmt`, due to technical debt that made it difficult to add proper error reporting in all situations.

The goal of SeoYoung's project was to perform a large internal refactoring of `rustfmt` that would allow tracking context about what went wrong during reformatting. In turn, this would enable turning silent failures into proper error messages that could help users examine and debug what went wrong. It could even allow `rustfmt` to retry formatting in more situations.

> TODO

Reviewer 8: @Kobzol Thank you for the writeup. 

As you already mentioned, the primary objective of the project was to refactor the codebase to track error contexts. We introduced a new trait to enhance Rewrite trait and modified previous implementations to align with this new trait. Now, rustfmt has specific information on where formatting failures occur and why they happen. Specifically, it has types of formatting errors it previously couldn’t report, including issues related to macro formatting logic. Additionally, rustfmt has ability to provide related span information, identifying which parts of the code require spacing adjustments on exceeding the maximum line width error.
For potential contributors examining the codebase, the clearer modeling of internal formatting failures offers some insights. They can better discern whether an issue is indeed a failure or simply a different expected formatting scenario that can be handled in the next attempt. (This part is my personal opinion, though.) 
Throughout this process, I also took the opportunity to remove outdated comments and enhance the overall codebase quality. 
However, I couldn’t achieve the optional project goal of enhancing user messages based on this error-related information. This means there’s still work to be done for users to experience the project’s full benefits directly. Nonetheless, during the project period, I already discussed the error propagation part with my mentor, identifying which parts of the codebase would need adjustments to propagate error messages to the user effectively. I also detailed any conflicts with rustfmt’s current error reporting style and other considerations. So the remaining task is to proceed with the work step by step according to the plan. I took some break but I plan to continue this work since I remember the satisfaction of displaying the error context information gathered through this large refactoring job during past meetings, where I tested it as a proof of concept. 

This project’s focus on internal codebase improvements, and my personal limitations that prevented me from achieving the second project goal made it hard to show direct results to users. Still, I hope my explanation effectively highlights the project’s achievements.

Reviewer 9: Right, the main objective of the project was focused on an internal code refactor that would help rustfmt preserve more context around formatting failures. At first blush this seemed like a fairly straightforward task. Just replace all formatting function return types to use `Result<String, RewriteError>` instead of `Option<String>`, but it quickly became clear that it wouldn't be that simple.

First, SeoYoung needed to come up with an approach to incrementally apply these refactors that would be easy to review and wouldn't impact the entire code base. Then while SeoYoung worked on certain sections of the codebase edge cases emerged that we hadn't anticipated when the project got started. SeoYoung was meticulous and systematic with her approach, and made sure that no formatting function or methods were missed.

Ultimately, the refactor was a success and internally rustfmt is keeping track of more information when formatting fails. We don't currently propagating that additional failure context as user facing error messages because that was a reach goal for the project that we didn't have time to complete, but SeoYoung has expressed interest in continuing to work on that as a future improvement!

Author: Thank you both! I tried to take your descriptions and combine them into a coherent text. Please let me know what do you think: https://github.com/rust-lang/blog.rust-lang.org/pull/1419/commits/7ab58022a4324d9cbdac25ea6cafedb77f686838

By the end of the GSoC project, the .NET backend was able to properly compile and run almost 90% of the standard library `core` and `std` test suite. That is an incredibly impressive number, since the suite contains thousands of tests, some of which are quite *arcane*. Michał's pace has not slowed down even after the project has ended and he is still continuously improving the backend. Oh, and did we already mention that his backend also has experimental support for emitting *C* code, effectively acting as a *C* codegen backend?! Michał has been very busy over the summer.

We thank Michał for all his work on the .NET backend, as it was truly inspirational, and led to fruitful discussions that were relevant also to other codegen backends. Michał's next goal is to get his backend upstreamed and create an official .NET compilation target, which could open up the doors to Rust becoming a first-class citizen in the .NET ecosystem. 

[codegen-clr]: https://github.com/FractalFir/rustc_codegen_clr

> ### Sandboxed and Deterministic Proc Macro using Wasm

Reviewer 7: ```suggestion
### Sandboxed and deterministic proc macro using WebAssembly
```

Author: These are official project names, but I guess it doesn't hurt to e.g. elaborate the abbreviations for the blog post. Made the word start with a lower case letter, aslo in the other cases. Thanks!

The goal of SeoYoung's project was to perform a large internal refactoring of `rustfmt` that would allow tracking context about what went wrong during reformatting. In turn, this would enable turning silent failures into proper error messages that could help users examine and debug what went wrong, and could even allow `rustfmt` to retry formatting in more situations.

At first, this might sound like an easy task, but performing such large-scale refactoring within a complex project such as `rustfmt` is not so simple. SeoYoung needed to come up with an approach to incrementally apply these refactors, so that they would be easy to review and wouldn't impact the entire code base at once. She introduced a new trait that enhanced the original `Rewrite` trait, and modified existing implementations to align with it. She also had to deal with various edge cases that we hadn't anticipated before the project started. SeoYoung was meticulous and systematic with her approach, and made sure that no formatting functions or methods were missed.

> Ultimately, the refactor was a success! Internally, rustfmt now keeps track of more information related to formatting failures, including errors that it could not possibly report before, such as issues with macro formatting. It also has the ability to provide information about source code spans, which helps identify parts of code that require spacing adjustments when exceeding the maximum line width. We don't yet currently propagate that additional failure context as user facing error messages, as that was a stretch goal that we didn't have time to complete, but SeoYoung has expressed interest in continuing to work on that as a future improvement!

Reviewer 9: @Kobzol The text here looks good! Just one minor suggestion:
```suggestion
Ultimately, the refactor was a success! Internally, rustfmt now keeps track of more information related to formatting failures, including errors that it could not possibly report before, such as issues with macro formatting. It also has the ability to provide information about source code spans, which helps identify parts of code that require spacing adjustments when exceeding the maximum line width. We don't yet propagate that additional failure context as user facing error messages, as that was a stretch goal that we didn't have time to complete, but SeoYoung has expressed interest in continuing to work on that as a future improvement!
```

Author: Thanks, modified it.

Apart from working on error context propagation, SeoYoung also made various other improvements that enhanced the overall quality of the codebase, and she was also helping other contributors understand `rustfmt`. Thank you for making the foundations of formatting better for everyone!

[rustfmt]: https://github.com/rust-lang/rustfmt

### Rust to .NET compiler - add support for compiling & running cargo tests

- Contributor: [Michał Kostrubiec](https://github.com/FractalFir)
- Mentor: [Jack Huey](https://github.com/jackh726)
- [Final report](https://fractalfir.github.io/generated_html/rustc_codegen_clr_v0_2_0.html)

> As was already mentioned above, the Rust compiler can be used with various codegen backends. One of these is the [.NET backend][codegen-clr], which compiles Rust code to the Common Intermediate Language (CIL), which can then be executed by the .NET Common Language Runtime (CRL). This backend allows interoperability of Rust and .NET (e.g. C#) code, in an effort to bring these two ecosystems closer together.

Author: Thanks!
